File: src/main/java/com/chatdemo/ChatApplication.java
package com.chatdemo;

import com.chatdemo.config.FirebaseConfig;
import com.chatdemo.config.ModelsConfig;
import com.chatdemo.memory.HistoryAwareChatMemoryStore;
import com.chatdemo.memory.SummarizingTokenWindowChatMemory;
import com.chatdemo.repository.ConversationRepository;
import com.chatdemo.repository.InMemoryConversationRepository;
import com.chatdemo.storage.FirebaseStorageUploader;
import com.chatdemo.tools.ImageGenerationTool;
import dev.langchain4j.memory.chat.ChatMemoryProvider;
import dev.langchain4j.model.TokenCountEstimator;
import dev.langchain4j.model.chat.ChatModel;
import dev.langchain4j.model.chat.StreamingChatModel;
import dev.langchain4j.model.image.DisabledImageModel;
import dev.langchain4j.model.image.ImageModel;
import dev.langchain4j.model.openai.OpenAiTokenCountEstimator;
import dev.langchain4j.service.AiServices;
import dev.langchain4j.service.MemoryId;
import dev.langchain4j.service.SystemMessage;
import dev.langchain4j.service.TokenStream;
import dev.langchain4j.service.UserMessage;

import java.util.List;
import java.util.Scanner;
import java.util.concurrent.CountDownLatch;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Main CLI application for multi-provider AI chatbot using LangChain4j.
 * Supports switching between Gemini, ChatGPT, Claude, and Grok.
 * Uses summarizing chat memory with full history preservation.
 */
public class ChatApplication {
    
    private static final String MEMORY_ID = "default";
    private static final int MAX_TOKENS = 4000;
    private static final String ANSI_RESET = "\u001B[0m";
    private static final String ANSI_USER = "\u001B[38;5;33m";
    private static final String ANSI_MODEL = "\u001B[38;5;214m";
    private static final Pattern URL_PATTERN = Pattern.compile("(https?://\\S+)");
    
    private final List<ProviderConfig> configs;
    private final ConversationRepository conversationRepository;
    private final HistoryAwareChatMemoryStore memoryStore;
    private final TokenCountEstimator tokenCountEstimator;
    private final ImageGenerationTool imageTool;
    private final String defaultOpenAiImageModel;
    private final String defaultGeminiImageModel;
    private final String defaultGrokImageModel;
    private int currentIndex;
    private Assistant currentAssistant;
    private String currentImageProvider;
    private String currentOpenAiImageModel;
    private String currentGeminiImageModel;
    private String currentGrokImageModel;
    private String lastImageUrl;
    
    /**
     * AI Assistant interface - LangChain4j will implement this.
     */
    interface Assistant {
        @SystemMessage("""
            You are a helpful assistant.
            If the user asks to create, draw, generate, or edit images, call the image tool.
            When an image URL is provided and the user wants modifications, use the edit tool.
            Return the tool output directly to the user.
            """)
        TokenStream chat(@MemoryId String memoryId, @UserMessage String message);
    }
    
    public ChatApplication() {
        this.configs = ModelsConfig.PROVIDERS;
        this.conversationRepository = new InMemoryConversationRepository();
        this.memoryStore = new HistoryAwareChatMemoryStore(conversationRepository);
        this.tokenCountEstimator = new OpenAiTokenCountEstimator("gpt-4");
        this.defaultOpenAiImageModel = ModelFactory.defaultOpenAiImageModel();
        this.defaultGeminiImageModel = ModelFactory.defaultGeminiImageModel();
        this.defaultGrokImageModel = ModelFactory.defaultGrokImageModel();
        this.currentOpenAiImageModel = defaultOpenAiImageModel;
        this.currentGeminiImageModel = defaultGeminiImageModel;
        this.currentGrokImageModel = defaultGrokImageModel;
        this.currentImageProvider = "openai";
        this.imageTool = new ImageGenerationTool();
        configureFirebaseStorage();
        this.currentIndex = 0;
        this.currentAssistant = createAssistant(configs.get(currentIndex));
    }
    
    private Assistant createAssistant(ProviderConfig config) {
        ChatModel model = ModelFactory.createModel(config);
        StreamingChatModel streamingModel = ModelFactory.createStreamingModel(config);
        syncImageProviderWithChatModel(config);
        if ("claude".equalsIgnoreCase(config.providerType())) {
            imageTool.disableImageTools("Image generation is not available when using Anthropic.");
        } else {
            imageTool.enableImageTools();
        }
        refreshImageModels();
        
        ChatMemoryProvider memoryProvider = memoryId -> SummarizingTokenWindowChatMemory.builder()
            .id(memoryId)
            .maxTokens(MAX_TOKENS)
            .tokenCountEstimator(tokenCountEstimator)
            .chatMemoryStore(memoryStore)
            .chatModel(model)
            .build();
        
        return AiServices.builder(Assistant.class)
            .streamingChatModel(streamingModel)
            .chatMemoryProvider(memoryProvider)
            .tools(imageTool)
            .build();
    }

    private void syncImageProviderWithChatModel(ProviderConfig config) {
        switch (config.providerType().toLowerCase()) {
            case "gemini" -> {
                currentImageProvider = "gemini";
                imageTool.setProvider(ImageGenerationTool.Provider.GEMINI);
            }
            case "chatgpt" -> {
                currentImageProvider = "openai";
                imageTool.setProvider(ImageGenerationTool.Provider.OPENAI);
            }
            case "grok" -> {
                currentImageProvider = "grok";
                imageTool.setProvider(ImageGenerationTool.Provider.GROK);
            }
            default -> {
                // Keep current image provider (e.g., Claude has no image model).
            }
        }
    }
    
    public void run() {
        Scanner scanner = new Scanner(System.in);
        
        printWelcome();
        
        while (true) {
            System.out.print("\n" + ANSI_USER + "You" + ANSI_RESET + "\n");
            String input = scanner.nextLine().trim();
            
            if (input.isEmpty()) {
                continue;
            }
            
            if (input.equalsIgnoreCase("/quit") || input.equalsIgnoreCase("/exit")) {
                System.out.println("Goodbye!");
                break;
            }
            
            if (input.equalsIgnoreCase("/setmodel")) {
                handleSetModel(scanner);
                continue;
            }
            
            if (input.equalsIgnoreCase("/clear")) {
                handleClear();
                continue;
            }
            
            if (input.equalsIgnoreCase("/help")) {
                printHelp();
                continue;
            }

            if (input.toLowerCase().startsWith("/imagemodel")) {
                handleImageModel(input);
                continue;
            }
            
            if (input.startsWith("/")) {
                System.out.println("Unknown command. Type /help for available commands.");
                continue;
            }
            
            handleChat(input);
        }
        
        scanner.close();
    }
    
    private void printWelcome() {
        System.out.println("===========================================");
        System.out.println("   AI Chatbot CLI - LangChain4j Demo");
        System.out.println("===========================================");
        System.out.println();
        printHelp();
        System.out.println();
        System.out.println("Current model: " + configs.get(currentIndex).getDisplayName());
    }
    
    private void printHelp() {
        System.out.println("Commands:");
        System.out.println("  /setmodel  - Switch AI model");
        System.out.println("  /imagemodel - Switch image model (openai|gemini|grok)");
        System.out.println("  /clear     - Clear conversation history");
        System.out.println("  /help      - Show this help");
        System.out.println("  /quit      - Exit application");
    }

    private void handleImageModel(String input) {
        String[] parts = input.trim().split("\\s+");
        if (parts.length == 1) {
            printImageModelStatus();
            return;
        }
        
        String provider = parts[1].toLowerCase();
        String modelName = (parts.length >= 3) ? parts[2] : null;
        
        switch (provider) {
            case "openai", "dall-e", "dalle" -> {
                if (modelName != null && !modelName.isBlank()) {
                    currentOpenAiImageModel = modelName;
                }
                currentImageProvider = "openai";
                imageTool.setProvider(ImageGenerationTool.Provider.OPENAI);
                refreshImageModels();
                System.out.println("Image model set to OpenAI (" + imageTool.getOpenAiModelName() + ")");
            }
            case "gemini" -> {
                if (modelName != null && !modelName.isBlank()) {
                    currentGeminiImageModel = modelName;
                }
                currentImageProvider = "gemini";
                imageTool.setProvider(ImageGenerationTool.Provider.GEMINI);
                refreshImageModels();
                System.out.println("Image model set to Gemini (" + imageTool.getGeminiModelName() + ")");
            }
            case "grok" -> {
                if (modelName != null && !modelName.isBlank()) {
                    currentGrokImageModel = modelName;
                }
                currentImageProvider = "grok";
                imageTool.setProvider(ImageGenerationTool.Provider.GROK);
                refreshImageModels();
                System.out.println("Image model set to Grok (" + imageTool.getGrokModelName() + ")");
            }
            default -> {
                System.out.println("Unknown image model provider. Use /imagemodel openai, /imagemodel gemini, or /imagemodel grok.");
                System.out.println("Examples:");
                System.out.println("  /imagemodel openai");
                System.out.println("  /imagemodel gemini");
                System.out.println("  /imagemodel grok");
                System.out.println("  /imagemodel gemini " + defaultGeminiImageModel);
            }
        }
    }
    
    private void printImageModelStatus() {
        System.out.println("Image model providers:");
        System.out.println("  OpenAI: " + currentOpenAiImageModel);
        System.out.println("  Gemini: " + currentGeminiImageModel);
        System.out.println("  Grok: " + currentGrokImageModel);
        System.out.println("Current image provider: " + currentImageProvider + " (" + imageTool.getCurrentModelName() + ")");
        System.out.println("Examples:");
        System.out.println("  /imagemodel openai");
        System.out.println("  /imagemodel gemini");
        System.out.println("  /imagemodel grok");
        System.out.println("  /imagemodel gemini " + defaultGeminiImageModel);
    }
    
    private void refreshImageModels() {
        String openAiKey = ModelsConfig.getOpenAiKey();
        ImageModel openAiModel = (openAiKey == null || openAiKey.isBlank())
            ? new DisabledImageModel()
            : ModelFactory.createImageModel(openAiKey, currentOpenAiImageModel);
        
        String geminiKey = ModelsConfig.getGeminiKey();
        ChatModel geminiModel = (geminiKey == null || geminiKey.isBlank())
            ? null
            : ModelFactory.createGeminiImageModel(geminiKey, currentGeminiImageModel);

        String grokKey = ModelsConfig.getGrokKey();
        ImageModel grokModel = (grokKey == null || grokKey.isBlank())
            ? new DisabledImageModel()
            : ModelFactory.createGrokImageModel(grokKey, currentGrokImageModel);
        
        imageTool.setOpenAiImageModel(openAiModel, currentOpenAiImageModel);
        imageTool.setGeminiImageModel(geminiModel, currentGeminiImageModel);
        imageTool.setGrokImageModel(grokModel, currentGrokImageModel);
    }
    
    private void handleSetModel(Scanner scanner) {
        System.out.println("\nAvailable models:");
        for (int i = 0; i < configs.size(); i++) {
            String marker = (i == currentIndex) ? " [current]" : "";
            System.out.println("  " + (i + 1) + ". " + configs.get(i).getDisplayName() + marker);
        }
        
        System.out.print("\nSelect model (1-" + configs.size() + "): ");
        String choice = scanner.nextLine().trim();
        
        try {
            int index = Integer.parseInt(choice) - 1;
            if (index >= 0 && index < configs.size()) {
                currentIndex = index;
                currentAssistant = createAssistant(configs.get(currentIndex));
                System.out.println("Switched to: " + configs.get(currentIndex).getDisplayName());
                System.out.println("(Conversation history preserved)");
            } else {
                System.out.println("Invalid selection. Please choose 1-" + configs.size());
            }
        } catch (NumberFormatException e) {
            System.out.println("Invalid input. Please enter a number.");
        }
    }
    
    private void handleClear() {
        // Clear the repository and recreate assistant
        memoryStore.deleteMessages(MEMORY_ID);
        currentAssistant = createAssistant(configs.get(currentIndex));
        System.out.println("Conversation history cleared. Starting fresh!");
    }
    
    private void handleChat(String message) {
        String providerName = configs.get(currentIndex).getDisplayName().split(" ")[0];
        CountDownLatch done = new CountDownLatch(1);
        System.out.println();
        System.out.print(ANSI_MODEL + providerName + ANSI_RESET + "\n");
        
        try {
            TokenStream stream = currentAssistant.chat(MEMORY_ID, message);
            stream.onPartialResponse(token -> {
                System.out.print(token);
                System.out.flush();
            }).onCompleteResponse(response -> {
                System.out.println();
                done.countDown();
            })
                .onError(error -> {
                    System.out.println();
                    System.out.println("[Error] " + error.getMessage());
                    done.countDown();
                })
                .start();
            done.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            System.out.println("[Error] Interrupted while waiting for response.");
        } catch (Exception e) {
            System.out.println("[Error] " + e.getMessage());
            done.countDown();
        }
    }

    private void configureFirebaseStorage() {
        FirebaseStorageUploader uploader = new FirebaseStorageUploader(
            FirebaseConfig.getServiceAccountPath(),
            FirebaseConfig.getStorageBucket()
        );
        if (uploader.isConfigured()) {
            imageTool.setFirebaseStorageUploader(uploader);
        }
    }

    private void handleImageRequest(String message) {
        String providerName = "Image model: " + imageTool.getCurrentModelName();
        System.out.println();
        System.out.print(ANSI_MODEL + providerName + ANSI_RESET + "\n");
        String response;
        String imageUrl = extractFirstUrl(message);
        if (imageUrl == null && isImageEditRequest(message)) {
            imageUrl = lastImageUrl;
        }
        if (imageUrl != null && isImageEditRequest(message)) {
            String prompt = stripUrl(message, imageUrl).trim();
            if (prompt.isBlank()) {
                prompt = "Edit the image as requested.";
            }
            response = imageTool.editImage(imageUrl, prompt);
        } else {
            response = imageTool.generateImage(message);
        }
        String responseUrl = extractFirstUrl(response);
        if (responseUrl != null) {
            lastImageUrl = responseUrl;
        }
        System.out.println(response);
    }

    private boolean isImageRequest(String message) {
        String lower = message.toLowerCase();
        return lower.contains("image")
            || lower.contains("draw")
            || lower.contains("sketch")
            || lower.contains("illustration")
            || lower.contains("picture")
            || lower.contains("photo")
            || lower.contains("generate")
            || lower.contains("create")
            || isImageEditRequest(lower);
    }

    private boolean isImageEditRequest(String message) {
        String lower = message.toLowerCase();
        return lower.contains("edit")
            || lower.contains("modify")
            || lower.contains("change")
            || lower.contains("transform")
            || lower.contains("add")
            || lower.contains("remove")
            || lower.contains("replace");
    }

    private String extractFirstUrl(String message) {
        Matcher matcher = URL_PATTERN.matcher(message);
        return matcher.find() ? matcher.group(1) : null;
    }

    private String stripUrl(String message, String url) {
        return message.replace(url, "");
    }
    
    public static void main(String[] args) {
        ChatApplication app = new ChatApplication();
        app.run();
    }
}

File: src/main/scala/com/chatdemo/app/ChatApplication.scala
package com.chatdemo.app

import com.chatdemo.common.model.{Conversation, ConversationMessage, MessageAttachment, MessageAttachmentExtractor, UserContext}
import com.chatdemo.common.service.ChatBackend

import java.util.{Locale, Scanner}
import scala.collection.mutable.ArrayBuffer

/**
 * Thin CLI client for the chat application.
 * Depends only on common interfaces - never imports backend.
 *
 * Model selection and pending attachments are kept client-side
 * and sent inline with each chat request (stateless server contract).
 */
class ChatApplication(private val backend: ChatBackend) {

  private val AnsiReset = "\u001B[0m"
  private val AnsiUser  = "\u001B[38;5;33m"
  private val AnsiModel = "\u001B[38;5;214m"
  private val AnsiDim   = "\u001B[2m"

  private val FreeUserContext = UserContext(isPremium = "false", deviceId = "device-free-001", signedInId = None)
  private val PremiumAppleUserContext = UserContext(isPremium = "true", deviceId = "device-premium-001", signedInId = Some("apple-user-001"))

  private var currentConversationId: String = _
  private var currentModelId: String = _
  private var currentUserContext: UserContext = FreeUserContext
  private val pendingAttachments: ArrayBuffer[MessageAttachment] = ArrayBuffer.empty

  def run(): Unit = {
    val scanner = new Scanner(System.in)

    printWelcome()

    var running = true
    while (running) {
      val prompt = if (currentConversationId != null) {
        s"\n${AnsiUser}You${AnsiDim} [$currentConversationId]${AnsiReset}\n"
      } else {
        s"\n${AnsiUser}You${AnsiReset}\n"
      }
      print(prompt)
      val input = normalizeInput(scanner.nextLine())

      if (input.isEmpty) {
        // skip
      } else if (input.equalsIgnoreCase("/quit") || input.equalsIgnoreCase("/exit")) {
        println("Goodbye!")
        running = false
      } else if (input.equalsIgnoreCase("/setmodel")) {
        handleSetModel(scanner)
      } else if (input.equalsIgnoreCase("/clear")) {
        handleClear()
      } else if (input.equalsIgnoreCase("/help")) {
        printHelp()
      } else if (input.equalsIgnoreCase("/c") || input.toLowerCase(Locale.ROOT).startsWith("/c ")) {
        handleConversationCommand(input)
      } else if (input.equalsIgnoreCase("/dump")) {
        handleDump()
      } else if (input.toLowerCase.startsWith("/attach")) {
        handleAttachCommand(input)
      } else if (input.toLowerCase.startsWith("/premium")) {
        handlePremium(input)
      } else if (input.startsWith("/")) {
        println("Unknown command. Type /help for available commands.")
      } else {
        handleChat(input)
      }
    }

    scanner.close()
  }

  private def printWelcome(): Unit = {
    println("===========================================")
    println("   AI Chatbot CLI - LangChain4j Demo")
    println("===========================================")
    println()
    printHelp()
    println()
    val models = backend.getAvailableModels
    if (models.nonEmpty) {
      if (currentModelId == null || !models.exists(_.modelId == currentModelId)) {
        currentModelId = models.head.modelId
      }
      println("Current model: " + models.find(_.modelId == currentModelId).map(_.getDisplayName).getOrElse(models.head.getDisplayName))
    }
    println()
    println("Use /c <id> to create or load a conversation, or /c to list all conversations.")
  }

  private def printHelp(): Unit = {
    println("Commands:")
    println("  /c <id>       - Create or load a conversation")
    println("  /c            - List all conversations")
    println("  /dump         - Dump current conversation messages")
    println("  /setmodel     - Switch AI model")
    println("  /premium      - Set premium mode (on|off)")
    println("  /attach <url> - Stage attachment for next message")
    println("  /clear        - Clear current conversation history")
    println("  /help         - Show this help")
    println("  /quit         - Exit application")
  }

  // ----------------------------------------------------------------
  // Conversation commands
  // ----------------------------------------------------------------

  private def handleConversationCommand(input: String): Unit = {
    val parts = input.trim.split("\\s+", 2)
    if (parts.length < 2 || parts(1).isBlank) {
      handleListConversations()
      return
    }
    val id = parts(1).trim
    val created = backend.createConversation(currentUserContext, id)
    currentConversationId = id
    if (created) {
      println("Created conversation: " + id)
    } else {
      val history = backend.getConversationHistory(currentUserContext, id)
      printConversationPreview(id, history)
    }
  }

  private def printConversationPreview(id: String, history: List[ConversationMessage]): Unit = {
    if (history.isEmpty) {
      println(s"Loaded conversation: $id (empty)")
    } else {
      println(s"Loaded conversation: $id (${history.size} messages)")
      val start = Math.max(0, history.size - 6)
      if (start > 0) {
        println(s"  ... ($start earlier messages)")
      }
      for (i <- start until history.size) {
        val msg = history(i)
        val role = messageRole(msg)
        val text = messageText(msg)
        val preview = if (text.length > 120) text.substring(0, 120) + "..." else text
        println(s"  [$role] $preview")
      }
    }
  }

  private def handleListConversations(): Unit = {
    val conversations = backend.listConversations(currentUserContext)
    if (conversations.isEmpty) {
      println("No conversations yet. Use /c <id> to start one.")
    } else {
      println("Conversations:")
      for (conversation <- conversations) {
        printConversationListItem(conversation)
      }
    }
  }

  private def handleDump(): Unit = {
    if (currentConversationId == null) {
      println("No active conversation. Use /c <id> first.")
      return
    }
    val history = backend.getConversationHistory(currentUserContext, currentConversationId)
    if (history.isEmpty) {
      println(s"Conversation '$currentConversationId' has no messages.")
      return
    }
    println(s"Dumping conversation: $currentConversationId (${history.size} messages)")
    for (msg <- history) {
      val role = messageRole(msg)
      val text = messageText(msg)
      println(s"[$role] $text")
      if (msg.attachments != null && msg.attachments.nonEmpty) {
        for (attachment <- msg.attachments) {
          val t = if (attachment.attachmentType == null) "link" else attachment.attachmentType.toLowerCase(Locale.ROOT)
          val mime = if (attachment.mimeType == null) "" else s" (${attachment.mimeType})"
          println(s"  -> attachment [$t] ${attachment.url}$mime")
        }
      }
    }
  }

  // ----------------------------------------------------------------
  // Model selection (client-side, sent per-request)
  // ----------------------------------------------------------------

  private def handleSetModel(scanner: Scanner): Unit = {
    val models = backend.getAvailableModels

    println("\nAvailable models:")
    for (i <- models.indices) {
      val marker = if (models(i).modelId == currentModelId) " [current]" else ""
      println(s"  ${i + 1}. ${models(i).getDisplayName}$marker")
    }

    print(s"\nSelect model (1-${models.size}): ")
    val choice = scanner.nextLine().trim

    try {
      val index = Integer.parseInt(choice) - 1
      if (index >= 0 && index < models.size) {
        currentModelId = models(index).modelId
        println("Switched to: " + models(index).getDisplayName)
        println("(Will be used on next chat request)")
      } else {
        println(s"Invalid selection. Please choose 1-${models.size}")
      }
    } catch {
      case _: NumberFormatException =>
        println("Invalid input. Please enter a number.")
    }
  }

  private def handleClear(): Unit = {
    if (currentConversationId == null) {
      println("No active conversation. Use /c <id> first.")
      return
    }
    backend.clearConversation(currentUserContext, currentConversationId)
    println("Conversation history cleared. Starting fresh!")
  }

  private def handleChat(message: String): Unit = {
    if (currentConversationId == null) {
      println("No active conversation. Use /c <id> first.")
      return
    }

    val models = backend.getAvailableModels
    if (models.nonEmpty && (currentModelId == null || !models.exists(_.modelId == currentModelId))) {
      currentModelId = models.head.modelId
    }
    val providerName = if (models.nonEmpty) {
      models.find(_.modelId == currentModelId).map(_.getDisplayName.split(" ")(0)).getOrElse("AI")
    } else {
      "AI"
    }
    println()
    print(s"${AnsiModel}$providerName${AnsiReset}\n")

    // Consume pending attachments for this request
    val attachments = pendingAttachments.toList
    pendingAttachments.clear()

    val resolvedModelId = if (currentModelId == null || currentModelId.isBlank) {
      models.headOption.map(_.modelId).getOrElse("")
    } else {
      currentModelId
    }
    val result = backend.chat(currentConversationId, message, attachments, resolvedModelId, None, currentUserContext,
      new com.chatdemo.common.service.ChatStreamHandler {
        override def onToken(token: String): Unit = {
          print(token)
          System.out.flush()
        }
      })

    println()

    if (result.error != null) {
      println("[Error] " + result.error)
    }

    printAttachments(result.responseAttachments)
  }

  private def handleAttachCommand(input: String): Unit = {
    if (currentConversationId == null) {
      println("No active conversation. Use /c <id> first.")
      return
    }
    val parts = input.trim.split("\\s+", 2)
    if (parts.length < 2 || parts(1).isBlank) {
      println("Usage: /attach <URL>")
      return
    }
    val attachment = MessageAttachmentExtractor.fromUrl(parts(1).trim)
    if (attachment == null) {
      println("Could not parse URL for attachment.")
      return
    }
    pendingAttachments.append(attachment)
    println(s"Attachment staged: [${attachment.attachmentType}] ${attachment.url}")
  }

  private def handlePremium(input: String): Unit = {
    val parts = input.trim.split("\\s+")
    if (parts.length == 1) {
      val status = if (currentUserContext.isPremium.toBoolean) "on (premium apple)" else "off (free)"
      println(s"Premium mode is currently: $status")
      println("Usage: /premium on|off")
      return
    }
    parts(1).toLowerCase match {
      case "on" | "true" | "1" =>
        currentUserContext = PremiumAppleUserContext
        println(s"Switched to premium apple user (${currentUserContext.effectiveId}).")
      case "off" | "false" | "0" =>
        currentUserContext = FreeUserContext
        println(s"Switched to free user (${currentUserContext.effectiveId}).")
      case _ =>
        println("Usage: /premium on|off")
    }
  }

  // ----------------------------------------------------------------
  // Helpers
  // ----------------------------------------------------------------

  private def printAttachments(attachments: List[MessageAttachment]): Unit = {
    if (attachments == null || attachments.isEmpty) {
      return
    }
    println("Attachments:")
    for (attachment <- attachments) {
      val t = if (attachment.attachmentType == null) "link" else attachment.attachmentType.toLowerCase(Locale.ROOT)
      val mime = if (attachment.mimeType == null) "" else s" (${attachment.mimeType})"
      println(s"- [$t] ${attachment.url}$mime")
    }
  }

  private def printConversationListItem(conversation: Conversation): Unit = {
    val marker = if (conversation.id == currentConversationId) " [active]" else ""
    val titlePart =
      if (conversation.title == null || conversation.title.isBlank || conversation.title == conversation.id) ""
      else s" - ${conversation.title}"
    println(s"  - ${conversation.id}$titlePart$marker")
  }

  private def messageRole(msg: ConversationMessage): String = {
    msg.message match {
      case _: dev.langchain4j.data.message.UserMessage   => "you"
      case _: dev.langchain4j.data.message.AiMessage     => "ai"
      case _: dev.langchain4j.data.message.SystemMessage  => "system"
      case _ => "?"
    }
  }

  private def messageText(msg: ConversationMessage): String = {
    msg.message match {
      case u: dev.langchain4j.data.message.UserMessage   => u.singleText()
      case a: dev.langchain4j.data.message.AiMessage     => a.text()
      case s: dev.langchain4j.data.message.SystemMessage  => s.text()
      case other => other.toString
    }
  }

  private def normalizeInput(raw: String): String = {
    if (raw == null) {
      return ""
    }
    val stripped = raw.strip
    val dePasted = stripped
      .replace("\u001b[200~", "") // bracketed-paste prefix used by some terminals
      .replace("\u001b[201~", "") // bracketed-paste suffix used by some terminals
      .strip
    val cleaned = dePasted.dropWhile(ch => Character.getType(ch) == Character.FORMAT)
    if (cleaned.isEmpty) {
      ""
    } else if (cleaned.startsWith("／") || cleaned.startsWith("⁄") || cleaned.startsWith("∕") || cleaned.startsWith("⧸") || cleaned.startsWith("╱")) {
      "/" + cleaned.substring(1)
    } else {
      cleaned
    }
  }
}
